File: BOJ/Stack/Kotlin/9935.kt
import java.util.*

fun main(args: Array<String>) = with(Scanner(System.`in`)) {
	var state_stack = Stack<Int>()
	
	val text = nextLine()
	val pattern = nextLine()

	var cursor = 0
	var result = StringBuilder()
	for (i in 0..text.length-1) {
		val ch = text[i]
		result.append(ch)
	
		val txt_size = result.length
		if (ch == pattern[cursor]) {
			cursor += 1	
		} else {
			if (ch == pattern[0]) {
				state_stack.push(cursor)
				cursor = 1
			} else {
				state_stack = Stack<Int>()
				cursor = 0
			}
		}

		if (cursor == pattern.length) {
			result.delete(txt_size - cursor, txt_size)
			cursor = if (state_stack.size > 0) state_stack.pop() else 0
		}
	}

	println(if (result.length == 0) "FRULA" else result.toString())
}

File: BOJ/Stack/Python/9935.py
target_string=input()
pattern=input()

result = ""
state_stack = []

pattern_length = len(pattern)
cursor = 0
for i in range(len(target_string)):
    result += target_string[i]
    if target_string[i] == pattern[cursor]:
        cursor += 1
    else:
        if target_string[i] == pattern[0]:
            state_stack.append(cursor)
            cursor = 1
        else:
            state_stack = []
            cursor = 0    

    if cursor == pattern_length:
        result = result[:-pattern_length]
        cursor = state_stack.pop() if len(state_stack) > 0 else 0
    
if result == "":
    print("FRULA")
else:
    print(result)
